### PR TITLE
FEAT: LOB detection and fetch strategy for FetchAll and FetchMany

### PR DESCRIPTION
**Pull request by gargsaumya**
### Work Item / Issue Reference  

> [AB#34162](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/34162)
[AB#38110](https://sqlclientdrivers.visualstudio.com/c6d89619-62de-46a0-8b46-70b92a84d85e/_workitems/edit/38110)

> GitHub Issue: #<ISSUE_NUMBER>

-------------------------------------------------------------------
### Summary   
  
This pull request improves the handling of large object (LOB) columns when fetching data from SQL Server in the `mssql_python/pybind/ddbc_bindings.cpp` file. The main change is to detect LOB columns (such as large strings or binary data) and switch to a per-row fetching strategy using `SQLGetData_wrap` for those columns, ensuring correct streaming and memory usage. For non-LOB columns, the batch fetching logic remains unchanged, but now includes logic to handle LOBs appropriately during batch fetches.

**LOB detection and fetch strategy:**

* Added logic in both `FetchMany_wrap` and `FetchAll_wrap` to detect LOB columns by checking column data types and sizes, and to fall back to per-row fetch using `SQLGetData_wrap` when LOBs are present. This avoids buffer overflows and streams LOBs correctly. [[1]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R2647-R2675) [[2]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R2769-R2797)

* Updated calls to `FetchBatchData` in both wrappers to pass the list of detected LOB columns, so batch fetches can handle them appropriately. [[1]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2677-R2690) [[2]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2770-R2812)

**Batch fetch improvements:**

* Modified the signature of `FetchBatchData` to accept the LOB columns list and updated its logic to handle LOB columns differently: instead of throwing exceptions when buffer sizes are insufficient, it now streams LOB data using `FetchLobColumnData`. [[1]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2345-R2345) [[2]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R2388-R2396) [[3]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1R2408-R2410) [[4]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2428-R2424) [[5]](diffhunk://#diff-dde2297345718ec449a14e7dff91b7bb2342b008ecc071f562233646d71144a1L2520-R2518)

### CODE REVIEW

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 1]
```diff
@@ -2753,6 +2765,35 @@ SQLRETURN FetchAll_wrap(SqlHandlePtr StatementHandle, py::list& rows) {
...
+        if ((dataType == SQL_WVARCHAR || dataType == SQL_WLONGVARCHAR || 
+             dataType == SQL_VARCHAR || dataType == SQL_LONGVARCHAR ||
+             dataType == SQL_VARBINARY || dataType == SQL_LONGVARBINARY) &&
+            (columnSize == 0 || columnSize == SQL_NO_TOTAL || columnSize > 8000)) {
+            lobColumns.push_back(i + 1); // 1-based
+        }
+    }
+
+    // If we have LOBs → fall back to row-by-row fetch + SQLGetData_wrap
+    if (!lobColumns.empty()) {
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:2875]: How are we handling error here? 

If `SQLFetch_ptr` or `SQLGetData_wrap` fails, the function returns immediately. Are we doing any cleanup of any resources allocated prior to or during the fetch loop (e.g., buffer allocations)?

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 2]
```diff
@@ -2660,6 +2643,35 @@ SQLRETURN FetchMany_wrap(SqlHandlePtr StatementHandle, py::list& rows, int fetch
...
+        LOG("LOB columns detected → using per-row SQLGetData path");
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:2754]: How are we handling error here? 

If `SQLFetch_ptr` or `SQLGetData_wrap` fails, the function returns immediately. Are we doing any cleanup of any resources allocated prior to or during the fetch loop (e.g., buffer allocations)?

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 3]
```diff
@@ -2660,6 +2643,35 @@ SQLRETURN FetchMany_wrap(SqlHandlePtr StatementHandle, py::list& rows, int fetch
         return ret;
     }
 
+    std::vector<SQLUSMALLINT> lobColumns;
+    for (SQLSMALLINT i = 0; i < numCols; i++) {
+        auto colMeta = columnNames[i].cast<py::dict>();
+        SQLSMALLINT dataType = colMeta["DataType"].cast<SQLSMALLINT>();
+        SQLULEN columnSize = colMeta["ColumnSize"].cast<SQLULEN>();
+
+        if ((dataType == SQL_WVARCHAR || dataType == SQL_WLONGVARCHAR ||
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:2744]: General Comment: Since this code is C++ and interacts with Python, ensure the logic in batch fetching and per-row fetching is thread-safe if used in multi-threaded scenarios (When Async is implemted). There is no explicit locking or GIL handling for Python objects in the fetch loop. We may want to keep this as a backlog to be attached to Async Feature

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 4]
```diff
@@ -2380,25 +2380,19 @@ SQLRETURN FetchBatchData(SQLHSTMT hStmt, ColumnBuffers& buffers, py::list& colum
...
                             numCharsInData));
                     } else {
-                        // In this case, buffer size is smaller, and data to be retrieved is longer
-                        // TODO: Revisit
-                        std::ostringstream oss;
-                        oss << "Buffer length for fetch (" << columnSize << ") is smaller, & data "
-                            << "to be retrieved is longer (" << numCharsInData << "). ColumnID - "
-                            << col << ", datatype - " << dataType;
-                        ThrowStdException(oss.str());
+                        row.append(FetchLobColumnData(hStmt, col, SQL_C_CHAR, false, false));
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:2487]: Previously, buffer overflows or size mismatches threw detailed exceptions describing the column, type, and expected size. With the new LOB handling (using `FetchLobColumnData`), these explicit error messages have been removed. Please ensure `FetchLobColumnData` raises meaningful exceptions (including column index, type, and error details) and logs errors appropriately.

Missing or unclear error propagation in LOB fetching can hide real bugs and make debugging difficult.

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 5]
```diff
@@ -2425,13 +2420,7 @@ SQLRETURN FetchBatchData(SQLHSTMT hStmt, ColumnBuffers& buffers, py::list& colum
...
 #endif
                     } else {
-                        // In this case, buffer size is smaller, and data to be retrieved is longer
-                        // TODO: Revisit
-                        std::ostringstream oss;
-                        oss << "Buffer length for fetch (" << columnSize << ") is smaller, & data "
-                            << "to be retrieved is longer (" << numCharsInData << "). ColumnID - "
-                            << col << ", datatype - " << dataType;
-                        ThrowStdException(oss.str());
+                        row.append(FetchLobColumnData(hStmt, col, SQL_C_WCHAR, true, false));
```
Comment (sumitmsft) [inline mssql_python/pybind/ddbc_bindings.cpp:2515]: Previously, buffer overflows or size mismatches threw detailed exceptions describing the column, type, and expected size. With the new LOB handling (using `FetchLobColumnData`), these explicit error messages have been removed. Please ensure `FetchLobColumnData` raises meaningful exceptions (including column index, type, and error details) and logs errors appropriately.

Missing or unclear error propagation in LOB fetching can hide real bugs and make debugging difficult.

File: tests/test_004_cursor.py [hunk 6]
```diff
@@ -5124,412 +5070,260 @@ def test_emoji_round_trip(cursor, db_connection):
...
-        cursor.execute("CREATE TABLE #pytest_varcharmax (col VARCHAR(MAX))")
+        cursor.execute("CREATE TABLE #pytest_nvarcharmax (col NVARCHAR(MAX))")
+        db_connection.commit()
+        cursor.execute("INSERT INTO #pytest_nvarcharmax VALUES (?)", [""])
         db_connection.commit()
 
-        db_connection.autocommit = False
-        rollback_str = "ROLLBACK" * 2000
-        cursor.execute("INSERT INTO #pytest_varcharmax VALUES (?)", [rollback_str])
-        db_connection.rollback()
```
Comment (sumitmsft) [inline tests/test_004_cursor.py]: Previous versions of the tests included scenarios where VARCHAR(MAX) and NVARCHAR(MAX) - (below) inserts were rolled back, confirming correct transactional behavior for LOB columns. These have been removed in the current PR. Any specific reason these were removed? 
Can we restore or add tests to ensure rollbacks work correctly for LOB data (VARCHAR(MAX), NVARCHAR(MAX), VARBINARY(MAX)), including edge cases. This is crucial for database integrity and ACID compliance.

**Review by sumitmsft — CHANGES_REQUESTED** [review]
Left few comments

File: mssql_python/pybind/ddbc_bindings.cpp [hunk 7]
```diff
@@ -2753,6 +2765,35 @@ SQLRETURN FetchAll_wrap(SqlHandlePtr StatementHandle, py::list& rows) {
...
+    std::vector<SQLUSMALLINT> lobColumns;
+    for (SQLSMALLINT i = 0; i < numCols; i++) {
+        auto colMeta = columnNames[i].cast<py::dict>();
+        SQLSMALLINT dataType = colMeta["DataType"].cast<SQLSMALLINT>();
+        SQLULEN columnSize = colMeta["ColumnSize"].cast<SQLULEN>();
+
+        if ((dataType == SQL_WVARCHAR || dataType == SQL_WLONGVARCHAR || 
+             dataType == SQL_VARCHAR || dataType == SQL_LONGVARCHAR ||
+             dataType == SQL_VARBINARY || dataType == SQL_LONGVARBINARY) &&
+            (columnSize == 0 || columnSize == SQL_NO_TOTAL || columnSize > 8000)) {
```
Comment (bewithgaurav) [inline mssql_python/pybind/ddbc_bindings.cpp]: can we place 8000 in a constant (`SQL_MAX_NON_LOB_SIZE` or `SQLSERVER_MAX_VARIABLE_COLUMN_LENGTH`)? we seem to use it in multiple places

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 1 above)
Comment (bewithgaurav) [reply mssql_python/pybind/ddbc_bindings.cpp:2875]: Buffer allocations (e.g., ColumnBuffers buffers(numCols, fetchSize);) are managed by C++ automatic scope-based destruction if they are stack-allocated objects. If an error occurs before buffers is allocated, there is nothing to clean up. For resources allocated inside the fetch loop (like temporary py::list row), these are managed by Python/C++ bindings and will be cleaned up by destructors or garbage collection.
`SQLGetData_wrap` contains error raising with itself.
@gargsaumya please feel free to add any other context if you have

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 2 above)
Comment (bewithgaurav) [reply mssql_python/pybind/ddbc_bindings.cpp:2754]: same response

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 4 above)
Comment (bewithgaurav) [reply mssql_python/pybind/ddbc_bindings.cpp:2487]: +1 to this

Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp:2487]: Added an additional log with these details, in case of SQL_ERROR in FetchLobColumnData

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 3 above)
Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp:2744]: Added this exercise in next semester 's worksheet.

File: mssql_python/pybind/ddbc_bindings.cpp (hunk 1 above)
Comment (gargsaumya) [reply mssql_python/pybind/ddbc_bindings.cpp:2875]: That's correct.
So, @sumitmsft - all temporary buffers used for LOB streaming inside FetchLobColumnData are scope-managed, so they are automatically freed if an exception occurs - no manual cleanup is needed.

File: tests/test_004_cursor.py (hunk 6 above)
Comment (gargsaumya) [reply tests/test_004_cursor.py]: yes, that was an outdated version, we still have it.
